File: 02.MeetingScheduler/meetingScheduler.ts
class Meeting {
    private _startTime: number;
    private _endTime: number;
    private _room: Room;
    constructor(startTime, endTime, room) {
        this._startTime = startTime;
        this._endTime = endTime;
        this._room = room;
    }
    get startTime(): number {
        return this._startTime;
    }
    get endTime(): number {
        return this._endTime;
    }
};

class Room {
    private _name: string;
    private calander: {[day: number]: Meeting[]};
    constructor(name: string) {
        this._name = name;
        this.calander = {};
    }
    get name(): string {
        return this._name;
    }
    book(day: number, start: number, end: number):boolean {
        if(this.calander[day] == undefined) 
            this.calander[day] = [];
        for(let meeting of this.calander[day]) {
            if(meeting.startTime < end && meeting.endTime > start)
                return false;
        }
        let m = new Meeting(start, end, this);
        this.calander[day].push(m);
        return true;
    }
};

class Scheduler {
    private rooms:Room[];
    constructor(rooms: Room[]) {
        this.rooms = rooms;
    }
    book(day: number, start: number, end: number): string {
        for(let room of this.rooms) {
            if(room.book(day, start, end))
                return room.name;
        }
        return "No rooms available at the moment!!";
    }
};


function main() {
    let room1 = new Room("Atlas");
    let room2 = new Room("Nexus");
    let room3 = new Room("HolyCow");

    let scheduler = new Scheduler([ room1, room2, room3]);

    console.log(scheduler.book(1, 2, 5));   // Atlas
    console.log(scheduler.book(1, 5, 8));   // Atlas
    console.log(scheduler.book(1, 4, 8));   // Nexus
    console.log(scheduler.book(1, 3, 6));   // HolyCow
    console.log(scheduler.book(1, 7, 8));   // HolyCow
    console.log(scheduler.book(2, 7, 8));   // No Available rooms at the moment
}

main();

File: 02.MeetingScheduler/meetingScheduler.cpp
#include<bits/stdc++.h>

using namespace std;

class Meeting;

/******************* Room Class *********************************/
typedef unordered_map<int, vector<Meeting>> Calander;
class Room {
    string name;
    Calander calander;
    public:
    Room(string name) {
        this->name = name;
    }
    const string getName();
    bool book(int day, int start, int end);
};

const string Room::getName() {
    return name;
}

/*****************************************************************/

/******************** Meeting **********************************/
class Meeting {
    public:
        Meeting(int start, int end, Room room):room(room) {
            this->start = start;
            this->end = end;
        }
        int getEnd();
        int getStart();
    private:
        int start, end;
        Room room;
};
int Meeting::getStart() {
    return start;
}
int Meeting::getEnd() {
    return end;
}
/****************************************************************/

bool Room::book(int day, int start, int end) {  // This function is kept seperate to access Meeting.
    for(Meeting &m: calander[day]) {
        if(m.getStart() < end && start < m.getEnd())
            return false;
    }
    Meeting meeting(start, end, *this);
    calander[day].push_back(meeting);

    return true;
}

/************************ Scheduler ******************************/
// This acts as an interface to book the rooms
class Scheduler {
    private:
    vector<Room> rooms;

    public:
    Scheduler(vector<Room> rooms) {
        this->rooms = rooms;
    }
    string book(int day, int start, int end) {   // This functions returns the name of the booked room.
        for(Room &room: rooms) {
            bool booked = room.book(day, start, end);
            if(booked) return room.getName();
        }
        return "No Available rooms at the moment!!";
    }
};
/*****************************************************************/

int main() {
    Room room1("Atlas");
    Room room2("Nexus");
    Room room3("HolyCow");

    Scheduler scheduler({ room1, room2, room3 });

    cout<<scheduler.book(1, 2, 5)<<endl;   // Atlas
    cout<<scheduler.book(1, 5, 8)<<endl;   // Atlas
    cout<<scheduler.book(1, 4, 8)<<endl;   // Nexus
    cout<<scheduler.book(1, 3, 6)<<endl;   // HolyCow
    cout<<scheduler.book(1, 7, 8)<<endl;   // HolyCow
    cout<<scheduler.book(1, 7, 8)<<endl;   // No Available rooms at the moment
    return 0;
}

